src/k8s/pkg/k8sd/app/cluster_util.go

```diff
@@ -0,0 +1,130 @@
+		}
+		if err := externalDatastoreCert.CheckCertificates(); err != nil {
+			return fmt.Errorf("failed to initialize cluster certificates: %w", err), nil, nil
+		}
+		if err := setup.EnsureExtDatastorePKI(snap, externalDatastoreCert); err != nil {
+			return fmt.Errorf("failed to write cluster certificates: %w", err), nil, nil
+		}
+		return nil, nil, externalDatastoreCert
+	default:
+		return fmt.Errorf("unsupported datastore %s, must be one of %v", cfg.APIServer.Datastore, setup.SupportedDatastores), nil, nil
+	}
+}
+
+func setupKubeconfigs(s *state.State, kubeConfigDir string, securePort int, caCert string) error {
+	// Generate kubeconfigs
+	for _, kubeconfig := range []struct {
+		file     string
+		username string
+		groups   []string
+	}{
+		{file: "admin.conf", username: "kubernetes-admin", groups: []string{"system:masters"}},
+		{file: "controller.conf", username: "system:kube-controller-manager"},
+		{file: "proxy.conf", username: "system:kube-proxy"},
+		{file: "scheduler.conf", username: "system:kube-scheduler"},
+		{file: "kubelet.conf", username: fmt.Sprintf("system:node:%s", s.Name()), groups: []string{"system:nodes"}},
+	} {
+		token, err := impl.GetOrCreateAuthToken(s.Context, s, kubeconfig.username, kubeconfig.groups)
+		if err != nil {
+			return fmt.Errorf("failed to generate token for username=%s groups=%v: %w", kubeconfig.username, kubeconfig.groups, err)
+		}
+		if err := setup.Kubeconfig(path.Join(kubeConfigDir, kubeconfig.file), token, fmt.Sprintf("127.0.0.1:%d", securePort), caCert); err != nil {
+			return fmt.Errorf("failed to write kubeconfig %s: %w", kubeconfig.file, err)
+		}
+	}
+	return nil
+
+}
+
+func setupControlPlaneServices(snap snap.Snap, s *state.State, cfg types.ClusterConfig, nodeIP net.IP) error {
+	// Configure services
+	if err := setup.Containerd(snap, nil); err != nil {
+		return fmt.Errorf("failed to configure containerd: %w", err)
+	}
+	if err := setup.KubeletControlPlane(snap, s.Name(), nodeIP, cfg.Kubelet.ClusterDNS, cfg.Kubelet.ClusterDomain, cfg.Kubelet.CloudProvider); err != nil {
+		return fmt.Errorf("failed to configure kubelet: %w", err)
+	}
+	if err := setup.KubeProxy(s.Context, snap, s.Name(), cfg.Network.PodCIDR); err != nil {
+		return fmt.Errorf("failed to configure kube-proxy: %w", err)
+	}
+	if err := setup.KubeControllerManager(snap); err != nil {
+		return fmt.Errorf("failed to configure kube-controller-manager: %w", err)
+	}
+	if err := setup.KubeScheduler(snap); err != nil {
+		return fmt.Errorf("failed to configure kube-scheduler: %w", err)
+	}
+	if err := setup.KubeAPIServer(snap, cfg.Network.ServiceCIDR, s.Address().Path("1.0", "kubernetes", "auth", "webhook").String(), true, cfg.APIServer.Datastore, cfg.APIServer.DatastoreURL, cfg.APIServer.AuthorizationMode); err != nil {
+		return fmt.Errorf("failed to configure kube-apiserver: %w", err)
+	}
+	return nil
+}
+
+func startServicesControlPlane(ctx context.Context, snap snap.Snap, datastore string) error {
```

Comment:
This function does not only start the control plane services, it also waits for kube-apiserver to become ready.